File: backend/src/main/kotlin/com/example/backend/dtos/ProduitDetailDto.kt
package com.example.backend.dtos

import lombok.Data
import java.math.BigDecimal
import java.time.LocalDateTime

data class ProduitRequestDto(
  val ean13: String?,
  val codebarre: String?,
  val codeLaborex: String?,
  val codeUbipharm: String?,
  val reference: String?,
  val nom: String?,
  val stock: Int?,
  val stockMax: Int?,
  val stockMin: Int?,
  val contenuDetail: String?,
  val prixDetail: Int?,
  val margeBeneficiaire: BigDecimal?,
  val tva: BigDecimal?,
  val etat: String?,
  val prixAchat: BigDecimal?,
  val prixVente: BigDecimal?,
  val reductionMax: Int?,
  val grossisteId: String?,
  val detailId: Int?,
  val categorie: Int?,
  val fournisseur: Int?,
  val forme: Int?,
  val fabriquant: Int?,
  val rayon: Int?,
  val etagere: String?,
  val magasin: Int?,
)

data class ProduitRequestNewDto(
  val id: Int?,
  val ean13: String?,
  val codebarre: String?,
  val codeLaborex: String?,
  val codeUbipharm: String?,
  val reference: String?,
  val nom: String?,
  val stock: Int?,
  val stockMax: Int?,
  val stockMin: Int?,
  val contenuDetail: String?,
  val prixDetail: String?,
  val produitDetail: String?,
  val etat: String?,
  val reductionMax: Int?,
  val detailId: Int?,
  val categorieId: Int?,
  val formeId: Int?,
  val fabriquantId: Int?,
  val rayonId: Int?,
  val etagere: String?,
  val magasinId: Int?,
)

data class ProduitResponseDto(
  var id: Int?,
  var nom: String,
  var description: String?,
  var codebarre: String?,
  var image: String?,
  var seuil: Int?,
  var categorieNom: String?,
  var uniteMesure: String?,
  var tva: BigDecimal?,
  var prixAchatInitial: BigDecimal?,
  var margeBeneficiaire: BigDecimal?,
  var prixVenteConseille: BigDecimal?,
  var prixVenteActuel: BigDecimal?,
  var quantiteTotaleEnStock: Int?,
  var dateCreation: LocalDateTime?,
  var dateModification: LocalDateTime?,
  var stockDetails: List<StockDetailDto>?
)

data class StockDetailDto(
  val enRayonId: Long?,
  val productNom: String?,
  val depotNom: String?,
  val rayonNom: String?,
  val quantite: Int,
  val datePeremption: LocalDateTime?,
  val numeroLot: String?
)

data class ProduitStockUpdateRequestDto(
  val quantiteChange: Int, // Positive for adding stock, negative for removing
  val depotId: Int,
  val rayonId: Int?,
  val numeroLot: String?, // Important for tracking specific batches
  val datePeremption: LocalDateTime? // For new stock entries
)

data class ProduitTarificationUpdateRequestDto(
  val nouveauPrixVente: BigDecimal,
  val dateDebut: LocalDateTime? // Defaults to now if null
)

// DTOs for Categorie and Fournisseur (simple for now)
data class CategorieDto(val id: Int?, val nom: String)
data class FournisseurDto(val id: Int?, val nom: String, val email: String?, val telephone: String?)
data class DepotDto(val id: Int?, val nom: String, val adresse: String?)
data class RayonDto(val id: Int?, val nom: String?, val code: String?)


data class ProduitDetailDto (
  val nom: String? = null,
  val reference: String? = null,
  val stock:Int? = 0,
  val stockMax:Int? = 0,
  val stockMin:Int? = 0,
  val prix: String? = null,
  val reductionMax:Int? = 0,
  val magasinId: Long? = null,
  val data: List<DataDto>? = null,
)

data class DataDto (
  val nom: String? = null,
  val contenuDetail: String? = null,
  val produitId: Long? = null
)

File: backend_py/pharmaPython/app/schemas/produit_detail_dto.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from decimal import Decimal

class ProduitRequestDto(BaseModel):
  ean13: Optional[str] = None
  codebarre: Optional[str] = None
  codeLaborex: Optional[str] = None
  codeUbipharm: Optional[str] = None
  reference: Optional[str] = None
  nom: Optional[str] = None
  stock: Optional[int] = None
  stockMax: Optional[int] = None
  stockMin: Optional[int] = None
  contenuDetail: Optional[str] = None
  prixDetail: Optional[int] = None
  margeBeneficiaire: Optional[Decimal] = None
  tva: Optional[Decimal] = None
  etat: Optional[str] = None
  prixAchat: Optional[Decimal] = None
  prixVente: Optional[Decimal] = None
  reductionMax: Optional[int] = None
  grossisteId: Optional[str] = None
  detailId: Optional[int] = None
  categorie: Optional[int] = None
  fournisseur: Optional[int] = None
  forme: Optional[int] = None
  fabriquant: Optional[int] = None
  rayon: Optional[int] = None
  etagere: Optional[str] = None
  magasin: Optional[int] = None

class ProduitRequestNewDto(BaseModel):
  id: Optional[int] = None
  ean13: Optional[str] = None
  codebarre: Optional[str] = None
  codeLaborex: Optional[str] = None
  codeUbipharm: Optional[str] = None
  reference: Optional[str] = None
  nom: Optional[str] = None
  stock: Optional[int] = 0
  stockMax: Optional[int] = 0
  stockMin: Optional[int] = 0
  contenuDetail: Optional[str] = None
  prixDetail: Optional[str] = None
  produitDetail: Optional[str] = None
  etat: Optional[str] = None
  reductionMax: Optional[int] = 0
  detailId: Optional[int] = None
  categorieId: Optional[int] = 0
  formeId: Optional[int] = 0
  fabriquantId: Optional[int] = 0
  rayonId: Optional[int] = 0
  etagere: Optional[str] = None
  magasinId: Optional[int] = 0

class StockDetailDto(BaseModel):
  enRayonId: Optional[int] = None
  productNom: Optional[str] = None
  depotNom: Optional[str] = None
  rayonNom: Optional[str] = None
  quantite: int
  datePeremption: Optional[datetime] = None
  numeroLot: Optional[str] = None

class ProduitResponseDto(BaseModel):
  id: Optional[int] = None
  nom: str
  description: Optional[str] = None
  codebarre: Optional[str] = None
  image: Optional[str] = None
  seuil: Optional[int] = None
  categorieNom: Optional[str] = None
  uniteMesure: Optional[str] = None
  tva: Optional[Decimal] = None
  prixAchatInitial: Optional[Decimal] = None
  margeBeneficiaire: Optional[Decimal] = None
  prixVenteConseille: Optional[Decimal] = None
  prixVenteActuel: Optional[Decimal] = None
  quantiteTotaleEnStock: Optional[int] = None
  dateCreation: Optional[datetime] = None
  dateModification: Optional[datetime] = None
  stockDetails: Optional[List[StockDetailDto]] = None

class ProduitResponseNewDto(BaseModel):
  nom: Optional[str] = None
  ean13: Optional[str] = None
  codeLaborex: Optional[str] = None
  codeUbipharm: Optional[str] = None
  reference: Optional[str] = None
  stock: Optional[int] = 0
  stockMax: Optional[int] = 0
  stockMin: Optional[int] = 0
  contenuDetail: Optional[str] = None
  prixDetail: Optional[str] = None
  etat: Optional[str] = None
  id: Optional[str] = None
  reductionMax: Optional[int] = 0
  detailId: Optional[str] = None
  categorieId: Optional[int] = 0
  formeId: Optional[int] = 0
  fabriquantId: Optional[int] = 0
  rayonId: Optional[int] = 0
  produitDetail: Optional[str] = None
  etagere: Optional[str] = None
  magasinId: Optional[int] = 0

class CategorieDto(BaseModel):
  id: Optional[int] = None
  nom: str

class FournisseurDto(BaseModel):
  id: Optional[int] = None
  nom: str
  email: Optional[str] = None
  telephone: Optional[str] = None

class DepotDto(BaseModel):
  id: Optional[int] = None
  nom: str
  adresse: Optional[str] = None

class RayonDto(BaseModel):
  id: Optional[int] = None
  nom: Optional[str] = None
  code: Optional[str] = None

class ProduitStockUpdateRequestDto(BaseModel):
  quantiteChange: int
  depotId: int
  rayonId: Optional[int] = None
  numeroLot: Optional[str] = None
  datePeremption: Optional[datetime] = None

class ProduitTarificationUpdateRequestDto(BaseModel):
  nouveauPrixVente: Decimal
  dateDebut: Optional[datetime] = None

class DataDto(BaseModel):
  nom: Optional[str] = None
  contenuDetail: Optional[str] = None
  produitId: Optional[int] = None

class ProduitDetailDto(BaseModel):
  nom: Optional[str] = None
  reference: Optional[str] = None
  stock: Optional[int] = 0
  stockMax: Optional[int] = 0
  stockMin: Optional[int] = 0
  prix: Optional[str] = None
  reductionMax: Optional[int] = 0
  magasinId: Optional[int] = None
  data: Optional[List[DataDto]] = None
